diff --git a/mods/ITEMS/mcl_maps/tgadec.lua b/mods/ITEMS/mcl_maps/tgadec.lua
index 95967ed2ad..0ca2ba8539 100644
--- a/mods/ITEMS/mcl_maps/tgadec.lua
+++ b/mods/ITEMS/mcl_maps/tgadec.lua
@@ -64,10 +64,15 @@ local function get_pixels_fmt_2 (pixels, width, height, data, offset)
 	end
 end
 
-local function convert (pixel)
-	local r = lshift (band (pixel, 0x7c00), 9)
-	local g = lshift (band (pixel, 0x3e0), 6)
-	local b = lshift (band (pixel, 0x1f), 3)
+local function expand5(x)
+	x = band(x, 0x1f)
+	return bor(lshift(x, 3), rshift(x, 2))
+end
+
+local function convert(pixel)
+	local r = lshift (expand5 (rshift (pixel, 10)), 16)
+	local g = lshift (expand5 (rshift (pixel, 5)), 8)
+	local b = expand5 (pixel)
 	local a = band (-rshift (pixel, 15), 0xff000000)
 	return bor (a, r, g, b)
 end
